samples/grpc/local-drone-control-scala/src/main/scala/charging/ChargingStation.scala

```diff
@@ -0,0 +1,282 @@
+          context.log.warn(
+            "Drone {} requested charging but is already charging. Ignoring.",
+            droneId)
+          Effect.none
+        } else if (state.dronesCharging.size >= state.chargingSlots) {
+          val earliestFreeSlot = state.dronesCharging.map(_.chargingDone).min
+          context.log.info(
+            "Drone {} requested charging but all stations busy, earliest free slot {}",
+            droneId,
+            earliestFreeSlot)
+          Effect.reply(replyTo)(
+            StatusReply.Success(AllSlotsBusy(earliestFreeSlot)))
+        } else {
+          // charge
+          val chargeCompletedBy =
+            Instant.now().plusSeconds(FullChargeTime.toSeconds)
+          context.log.info(
+            "Drone {} requested charging, will complete charging at {}",
+            droneId,
+            chargeCompletedBy)
+          val event = ChargingStarted(droneId, chargeCompletedBy)
+          Effect.persist(event).thenRun { (_: Option[State]) =>
+            timers.startSingleTimer(
+              CompleteCharging(droneId),
+              durationUntil(chargeCompletedBy))
+            // Note: The event is also the reply
+            replyTo ! StatusReply.Success(event)
+          }
+        }
+
+      case CompleteCharging(droneId) =>
+        if (state.dronesCharging.exists(_.droneId == droneId)) {
+          context.log.info("Drone {} completed charging", droneId)
+          Effect.persist(ChargingCompleted(droneId))
+        } else {
+          Effect.none
+        }
+
+      case GetState(replyTo) =>
+        Effect.reply(replyTo)(StatusReply.Success(state))
+
+    }
+  }
+
+  private def handleEvent(state: Option[State], event: Event): Option[State] = {
+    state match {
+      case None =>
+        event match {
+          case Created(locationId, chargingSlots) =>
+            Some(State(chargingSlots, Set.empty, locationId))
+          case unexpected =>
+            throw new IllegalArgumentException(
+              s"Got unexpected event ${unexpected} for uninitialized state")
+        }
+
+      case Some(state) =>
+        event match {
+          case Created(_, _) =>
+            context.log.warn("Saw a second created event, ignoring")
+            Some(state)
+          case ChargingStarted(droneId, chargeComplete) =>
+            Some(
+              state.copy(dronesCharging = state.dronesCharging + ChargingDrone(
+                droneId,
+                chargeComplete,
+                replicationContext.origin.id)))
+          case ChargingCompleted(droneId) =>
+            Some(
+              state.copy(dronesCharging =
+                state.dronesCharging.filterNot(_.droneId == droneId)))
+        }
+
+    }
+  }
+
+  private def handleRecoveryCompleted(state: State): Unit = {
+    // FIXME this scheme is not quite reliable, because station is not remembered/restarted
+    //       until next new event/command if edge system is shut down/restarted
```

Comment:
@patriknw this is the part I mentioned, one simple improvement could perhaps be to schedule/complete on both edge and cloud replicas instead of only edge/initiating node, and make completion idempotent.

Reply:
I haven't looked at the details here yet, but do we need to have the timer? Could it instead be another grpc request telling that drone charging is complete? Maybe that is somewhat more realistic? The drone would report when it is fully charged?

Reply:
Yeah, maybe that is even better.

Reply:
Changed to be up to drone requesting completion now 👍